pytometry/tools/_humap.py

```diff
@@ -0,0 +1,439 @@
+    erg = reached_lm / np.sum(reached_lm.data)
+    return csr_matrix(erg)
+
+def _get_landmarks(T, settings):
+    '''
+    Parameters
+    ----------
+    T
+       Transition matrix
+    settings
+       settings dict
+
+    Returns
+       list of landmark indices
+    -------
+    '''
+    n_events = T.shape[0]
+    proposals = np.zeros(n_events)  # counts how many times point has been reached
+    landmarks = list()  # list of landmarks
+    global HELPER_VAR
+    HELPER_VAR = {'T': T,
+                  'theta': settings['theta'],
+                  'beta': settings['beta'],
+                  'beta_thresh': settings['beta_thresh'],
+                  'n_events': n_events}
+    init_states = csr_matrix((np.ones(n_events), (range(n_events), range(n_events))))
+    hit_list = map(_helper_method_get_landmarks, [state for state in init_states])
+    # evaluate results
+    for state_hits in hit_list:  # for every states hit_list
+        for h in state_hits:  # for every hit in some states hit_list
+            proposals[h[0]] += h[1]
+
+    # collect landmarks
+    min_beta = settings['beta'] * settings['beta_thresh']
+    for prop in enumerate(proposals):
+        # if event has been hit min_beta times, it counts as landmark
+        if prop[1] > min_beta:
+            landmarks.append(prop[0])
+    return landmarks
+
+def _helper_method_get_landmarks(state):
+    ### Multiprocessing helper method ###
+    for i in range(HELPER_VAR['theta']):
+        state *= HELPER_VAR['T']
+    destinations = np.random.choice(range(HELPER_VAR['n_events']), HELPER_VAR['beta'], p=state.toarray()[0])
+    hits = np.zeros((HELPER_VAR['n_events']))
+    for d in destinations:
+        hits[d] += 1
+    return [(h[0], h[1]) for h in enumerate(hits) if h[1] > 0]
+
+def _calc_first_T(distances_nn, dim):
+    '''
+    Parameters
+    ----------
+    distances_nn
+       distances of nearest neighbor graph
+    dim
+       dimensions of Transition matrix
+
+    Returns
+       Transition matrix
+    -------
+    '''
+    probs = map(_helper_method_calc_T, [dist.data for dist in distances_nn])
+
+    data = []
+    for pr in probs:
+        data.extend(pr)
+    T = csr_matrix((data, distances_nn.indices, distances_nn.indptr), shape=(dim,dim))
+    return T
+
+
+def _helper_method_calc_T(dist):
+    ### Multiprocessing helmper method ###
+    d = dist / np.max(dist)
+    return softmax((-d ** 2) / _binary_search_sigma(d, len(d)))
+
+def _binary_search_sigma(d, n_neigh):
+    # binary search
+    sigma = 10  # Start Sigma
+    goal = np.log(n_neigh)  # log(k) with k being n_neighbors
+    # Do binary search until entropy ~== log(k)
+    while True:
+        ent = entropy(softmax((-d ** 2) / sigma))
+        # check sigma
+        if np.isclose(ent, goal):
+            return sigma
+        if ent > goal:
+            sigma *= 0.5
+        else:
+            sigma /= 0.5
+
+
+############################################################################
```

Comment:
If the Lasso Selector part is not finished yet, remove it from this PR and do a different PR for it.